# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
#lambda >input< x : what we want to do
# lets take a look at how we can create a function to add two to a variable
def add_two(x):
    return x+2

#then we can test it:
print(add_two(4))

# this is great if we are looking to reuse this function over nd over again 
# but what if tthis was something i only wanted to do once?
# lambda comes into play 
a_var = 4
print((lambda x: x+2)(a_var))


6
6


#### Saving to a Variable

In [2]:
# we can save a lambda function to a variable for reuseability 
# do not recommend this as you can write a function to do the same thing 
f_test = lambda x: x+2
f_test(4)

6

#### Multiple Inputs

In [3]:
# no variables
print((lambda x,y,z: x * y * z)(3,5,8))

x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [4]:
def multiply(f, num):
    '''
    f expects to see a function (lambda function)
    num expects a single number
    '''
    return f(num)
multiply(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [5]:
def multiply_test(num):
    return num * 4

#lets try func of another func
def return_function():
    test = 4
    def multiply(num):
        return num*2
    return multiply
f_return = return_function()
print(f_return(4))

# what if i wanted to instead return a lambda function from my regular function?
def return_lamb(b,c):
    return lambda x,a: x+a+b+c

r_lamb = return_lamb(4,5)
print(r_lamb)
print(r_lamb(5,5))


8
<function return_lamb.<locals>.<lambda> at 0x0000016BC8169BD0>
19


#### If Statements within Lambdas

In [6]:
# lambdas are set up to take in conditionla statements 
#lambda x: True if (condition happens) else False
# lambda x: do something if (condition happens) else do something else
f_conditional = lambda num: num *2 if num > 10 else num + 2
print(f_conditional(8))
print(f_conditional(20))

10
40


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [7]:
f = lambda x: x ** 3
result = f(2) 
print(result)

8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [8]:
#%time = magic


numbers = [4,11,20,3,15,20,11,2,3,6,5,1,44,1]
def squared_normal(val):
    return val ** 2
lst =[]
for num in numbers:
    squared_normal(num)



In [9]:
# by using a map function, we can iterate through a list and apply a function without ever having to iterate ourselves
''' 
map(function, iterable)
and iterable is something we can loop over: suck as a list, dict, tuple, set or a string.
normally, yuo can use a map with a predefined function, but we can use it with a lambda statement
'''
%time
def squared_map(num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2

numbers = [4,11,20,3,15,20,11,2,3,6,5,1,44,1]
more_nums = [4,10,3,2,6,10,15,3,4,7,6,0,41,2]



squared_numbers_mapped = list(map(squared_map, numbers, more_nums))
print(squared_numbers_mapped)

another_squared_nums_mapped = list(map(squared_normal, numbers))
print(another_squared_nums_mapped)



CPU times: total: 0 ns
Wall time: 0 ns
[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6), (20, 10), (11, 15), (4, 9), (9, 16), (36, 49), (25, 36), (1, 0), (44, 41), (1, 4)]
[16, 121, 400, 9, 225, 400, 121, 4, 9, 36, 25, 1, 1936, 1]


#### Using Lambda's with Map

In [11]:
list(map(lambda x,y: (x**2, y**2)if x<10 and y<10 else (x,y), numbers, more_nums))

[(16, 16),
 (11, 10),
 (20, 3),
 (9, 4),
 (15, 6),
 (20, 10),
 (11, 15),
 (4, 9),
 (9, 16),
 (36, 49),
 (25, 36),
 (1, 0),
 (44, 41),
 (1, 4)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [10]:
l_m = [20,3,10,6,3,12]
l_n = [30,2,15,4,5,16]

n_list= list(map(lambda x: x*2-1,l_m))
print(n_list)
picture = [ '','abc', 'ded', '']
list(map(lambda x: '*' + x + '*' if len(x) > 1 else '*****', picture))


[39, 5, 19, 11, 5, 23]


['*****', '*abc*', '*ded*', '*****']

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [11]:
# syntax for this looks like:
# list(filter(function, iterable(s)))

names = ['Bob', 'Andy', 'Max', 'Evan', 'Angela']
def a_names(name):
    #in this function, we will check to see if the first letter of each name begins w "A"
    if name[0].lower() == 'a':
        return True 
    else: # if it is not an "a"
        return False
list(filter(a_names, names))

['Andy', 'Angela']

#### Using Lambda's with Filter()

In [12]:
# we can also use Lambdas w the filter function!
# this is great if you are looking to filter data w a single-use function instead of having to define
# an create another function like we did above 
list(filter(lambda name: True if name[0].lower() == 'a' else False, names))

['Andy', 'Angela']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [13]:
import statistics
import numpy as np

# Using the statstics module to gather the mean
print(statistics.mean(numbers))

# Using the numpy library to gather the mean as well
print(np.mean(numbers))

# There are three ways to go about this.

# Number 1: Using vanilla Python
print(list(filter(lambda x: True if x>= sum(numbers)/ len(numbers) else False, numbers)))

# Number 2: Using the numpy library
print(list(filter(lambda x: True if x >= np.mean(numbers) else False, numbers)))

# Number 3: Using the Statistics module
print(list(filter(lambda x: True if x >= statistics.mean(numbers) else False, numbers)))


10.428571428571429
10.428571428571429
[11, 20, 15, 20, 11, 44]
[11, 20, 15, 20, 11, 44]
[11, 20, 15, 20, 11, 44]


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [2]:
### Show a way to break the program by creating an infinite loop


def add_nums_breakable(num):
    #without a base case, we willl end up infinetely looping though our program 
    if num >-20000:
        return num + add_nums_breakable(num-1)
    else:
        return num

add_nums_breakable(5)


: 

: 

In [1]:
'''what is base case?
 a base is what cause the function to break out of its infinite loop!
 it evaluates ithe value of the number or string at the moment it recutrs. to stop from going any further 
 these need to be implemented at the bekgining of your function, not after the recursive call
'''

def add_nums(num):
    if num<=1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums{num}) = {num} - add_nums({num - 1})')
        return num - add_nums(num-1)
add_nums(5)

add_nums5) = 5 - add_nums(4)
add_nums4) = 4 - add_nums(3)
add_nums3) = 3 - add_nums(2)
add_nums2) = 2 - add_nums(1)
add_nums(1) = 1


3

#### Writing a Factorial Function

In [2]:
#5! = 5 * 4 *3 * 2 * 1
def factorial(num):
    if num <= 1:
        print(f'factorial({num}) = {num}')
        return 1
    else:
        print(f'factorial({num}) = {num} * factorial({num - 1})')
print(factorial(5))
5*4*3*2*1

factorial(5) = 5 * factorial(4)
None


120

#### In-Class Exercise #4 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [3]:
def minus(num):
    if num <=1:
        print(f'minus_nums(num) = {num}')
        return num
    else:
        print(f'minus_nums{num}) = {num} - nums_nums({num - 1})')
        return num - minus(num-1)
minus(8)

minus_nums8) = 8 - nums_nums(7)
minus_nums7) = 7 - nums_nums(6)
minus_nums6) = 6 - nums_nums(5)
minus_nums5) = 5 - nums_nums(4)
minus_nums4) = 4 - nums_nums(3)
minus_nums3) = 3 - nums_nums(2)
minus_nums2) = 2 - nums_nums(1)
minus_nums(num) = 1


4

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [4]:
#print(type(range(0,10)))

# can be thought of as a conveyer belt 
def my_range(stop, start = 0, step = 2):
    while start < stop:
        yield start # yield  means that is going to be a generator. it will not return anything when the function is run 
        start += step
[i for i in my_range(10)]




[0, 2, 4, 6, 8]

#### In-Class Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [5]:
def squared_nums(num):
    while num > 0:
        yield num ** 2
        num -= 1
        
[i for i in squared_nums(5)]

[25, 16, 9, 4, 1]

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [8]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

#print("Original list is : " + str(places))
result = list(filter(lambda x: x.strip() != '', places))
print(result)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [10]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda x: x.split()[-1])
print(author)

['Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'Victor aNisimov', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [11]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

result = list(map(lambda x: (x[0], (9/5)*x[1] +32), places))
print(result)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [12]:
def fib_sequence(n):
    if n <= 0:
        return []
    elif n == 1:
        return [1]
    elif n == 2:
        return [1, 1]
    else:
        sequence = fib_sequence(n - 1)
        sequence.append(sequence[-1] + sequence[-2])
        return sequence


n = 5
result = fib_sequence(n)
for i in range(len(result)):
    print(f"Iteration {i}: {result[i]}")

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
